proxy/src/telemetry/metrics/serve.rs

```diff
@@ -63,22 +76,65 @@ impl Service for Serve {
             let mut writer = GzEncoder::new(Vec::<u8>::new(), CompressionOptions::fast());
             write!(&mut writer, "{}", *metrics)
                 .and_then(|_| writer.finish())
-                .map(|body| {
-                    Response::new()
-                        .with_header(ContentEncoding(vec![Encoding::Gzip]))
-                        .with_header(ContentType::plaintext())
-                        .with_body(Body::from(body))
+                .map_err(ServeError::from)
+                .and_then(|body| {
+                    Response::builder()
+                        .header(header::CONTENT_ENCODING, "gzip")
+                        .header(header::CONTENT_TYPE, "text/plain")
+                        .body(Body::from(body))
+                        .map_err(ServeError::from)
                 })
         } else {
             let mut writer = Vec::<u8>::new();
             write!(&mut writer, "{}", *metrics)
-                .map(|_| {
-                    Response::new()
-                        .with_header(ContentType::plaintext())
-                        .with_body(Body::from(writer))
+                .map_err(ServeError::from)
+                .and_then(|_| {
+                    Response::builder()
+                        .header(header::CONTENT_TYPE, "text/plain")
+                        .body(Body::from(writer))
+                        .map_err(ServeError::from)
                 })
         };
 
-        future::result(resp.map_err(hyper::Error::Io))
+        future::result(resp)
+    }
+}
+
+// ===== impl ServeError =====
+
+impl From<http::Error> for ServeError {
+    fn from(err: http::Error) -> Self {
+        ServeError::Http(err)
+    }
+}
+
+impl From<io::Error> for ServeError {
+    fn from(err: io::Error) -> Self {
+        ServeError::Io(err)
+    }
+}
+
+impl fmt::Display for ServeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}: {}",
+            self.description(),
+            self.cause().expect("ServeError must have cause")
+        )
+    }
+}
+
+impl Error for ServeError {
+    fn description(&self) -> &str {
+        match *self {
+            ServeError::Http(_) => "error constructing HTTP response",
```

Comment:
Note that an `Error` returned by the service will abort the connection. In 99% of cases, we probably wouldn't want to do that, and instead would want to return a response with a better status code. This error case likely should not be one of those to close the connection :)

Reply:
Changed in 86efd4f6b6b10a357bc5204fa5d43a8b7670c57c.